File: tau.pj/tests.t/cc.m/buf.cc
// Copyright (c) 2010 The Chromium OS Authors. All rights reserved.
// Distributed under the terms of the GNU General Public License v2

#include <sys/types.h>
#include <sys/uio.h>
#include <unistd.h>
#include <fcntl.h>

#include <iostream>
using namespace std;

#include <test.h>
#include <debug.h>
#include <eprintf.h>

enum { BUF_SIZE = 256 };

struct Buf_s {
  char b_data[BUF_SIZE];
  int  b_key;
  int  b_inuse;
};

class Dev {
  string name_;
  int fd_;
 public:
  explicit Dev(string filename);
  ~Dev();
  void fill(Buf_s *b);
  void flush(Buf_s *b);
};

Dev::Dev(string filename) {
  name_ = filename;
  fd_ = open(name_.c_str(), O_CREAT|O_RDWR, 0766);
}

Dev::~Dev() {
  close(fd_);
}

void Dev::fill(Buf_s *b) {
  int rc;
  rc = pread(fd_, b->b_data, BUF_SIZE, b->b_key * BUF_SIZE);
  if (rc < 0) fatal("pread:");
}

void Dev::flush(Buf_s *b) {
  int rc;
  rc = pwrite(fd_, b->b_data, BUF_SIZE, b->b_key * BUF_SIZE);
  if (rc < 0) fatal("pwrite:");
}

class Cache {
  Buf_s *bufs_;
  int   clock_;
  int   num_bufs_;
  Dev   dev_;
  Buf_s *lookup(int key);
  Buf_s *victim();
 public:
  Cache(string filename, int num_bufs);
  ~Cache();
  Buf_s *get_buf(int key);
  void   put_buf(Buf_s *b);
  void   clear();
  int    num_bufs() const { return num_bufs_; }
  bool   audit();
};

Cache::Cache(string filename, int num_bufs)
    : clock_(0),             // have to be initialized inorder of declaration
      num_bufs_(num_bufs),
      dev_(filename) {
  bufs_ = new Buf_s[num_bufs_];
  clear();
}

Cache::~Cache() {
  delete[] bufs_;
}

Buf_s *Cache::lookup(int key) {
  for (int i = 0; i < num_bufs_; ++i) {
    if (key == bufs_[i].b_key) return &bufs_[i];
  }
  return NULL;
}

Buf_s *Cache::victim() {
  Buf_s *b;
  do {
    ++clock_;
    if (clock_ >= num_bufs_) {
      clock_ = 0;
    }
    b = &bufs_[clock_];
  } while (b->b_inuse);
  return b;
}

Buf_s *Cache::get_buf(int key) {
  Buf_s *b = lookup(key);
  if (!b) {
    b = victim();
    b->b_key = key;
    dev_.fill(b);
  }
  ++b->b_inuse;
  return b;
}

void Cache::put_buf(Buf_s *b) {
  dev_.flush(b);
  --b->b_inuse;
}

void Cache::clear() {
  for (int i = 0; i < num_bufs_; ++i) {
    bufs_[i].b_key   = -1;
    bufs_[i].b_inuse = 0;
  }
}

bool Cache::audit() {
  if (clock_ >= num_bufs_) return false;
  return true;
}

int main(int argc, char *argv[]) {
  string cachefile = "/tmp/cache";
  int num_bufs = 10;
  if (argc > 1) {
    cachefile = argv[1];
  }
  Cache c(cachefile, num_bufs);
  Buf_s *b;

  // Test 1. Get a buffer, set a value, and put it back
  b = c.get_buf(1);
  test(b != NULL);
  test(b->b_key == 1);
  b->b_data[0] = 'A';
  c.put_buf(b);

  c.clear();

  // Test 2. Get the buffer from test 1 and check it has what we set
  // This is not a good test, because it depends on test 1.
  // This test did fail to detect that we read the block into a
  // separate buffer. Now we have to copies of the buffer in memory.
  b = c.get_buf(1);
  test(b != NULL);
  test(b->b_key == 1);
  test(b->b_data[0] == 'A');
  c.put_buf(b);
  test(c.audit());

  // Test 3. Use up all the buffers without releasing any of them
  // Can't really test for running off of end nor can we test for
  // infinite loops.
  for (int i = 0; i < c.num_bufs(); ++i) {
    b = c.get_buf(i);
    c.put_buf(b);
  }
  b = c.get_buf(c.num_bufs());
  test(b != NULL);

  test(c.audit());

  return 0;
}

File: tau.pj/go.t/go.m/buf.go
// Copyright 2010 Google Inc. All Rights reserved
// Author: taysom@google.com (Paul Taysom)

package main

import (
	"fmt"
	"runtime"
	"flag"
	"os"
	"unsafe"
)

func test(e bool) {
	if (e) { return }
	_, file, line, _ := runtime.Caller(1);
	fmt.Printf("Error at %s<%d>\n", file, line)
}

const BufSize = 256
const RootBlkno = 1
const RootMagic = 0x123456789abcdef

type Root struct {
	magic	int64
	nextblk	int64
}

type Buf struct {
	data	[BufSize]byte
	blkno	int64
	inuse	int
	cache	*Cache
}

type Dev struct {
	file	*os.File
	next	int64		// Next block num to allocate
}

type Cache struct {
	buf	[]Buf
	clock	int
	dev	*Dev
}

func OpenDev(name string) *Dev {
	f, e := os.Open(name, os.O_RDWR | os.O_CREAT, 0666);
	if e != nil {
		fmt.Fprintln(os.Stderr, "Failed to open", name, e)
		return nil
	}
	return &Dev{f, 1}
}

func (dev *Dev) NewBlock(b *Buf) {
	b.blkno = dev.next; dev.next++
	for i := 0; i < len(b.data); i++ {
		b.data[i] = 0
	}
}

func (dev *Dev) Fill(b *Buf) {
	_, e := dev.file.ReadAt(b.data[:], b.blkno * BufSize)
	if e != nil {
		if e.String() != "EOF" {
			fmt.Fprintln(os.Stderr, "Read failed", e)
		}
	}
}

func (dev *Dev) Flush(b *Buf) {
	_, e := dev.file.WriteAt(b.data[:], b.blkno * BufSize)
	if e != nil {
		fmt.Fprintln(os.Stderr, "Write failed", e)
	}
}

func (cache *Cache) readRoot() {
//	b := make([]byte, 256)
//	c := &b
	type rootptr *Root

	b := cache.GetBuf(RootBlkno)
	p := unsafe.Pointer(&b.data)
	r := rootptr(p)
fmt.Println("Root=", r, *r)
}

func NewCache(num_bufs int, dev *Dev) *Cache {
	b := make([]Buf, num_bufs)
	cache := &Cache{b, 0, dev}
	for i := 0; i < len(cache.buf); i++ {
		cache.buf[i].cache = cache
	}
	return cache
}

func (cache *Cache) Clear() {
	for i := 0; i < len(cache.buf); i++ {
		cache.buf[i].blkno = -1
		cache.buf[i].inuse = 0;
	}
}

func (cache *Cache) Audit() bool {
	if cache.clock >= len(cache.buf) { return false }
	return true
}

func (cache *Cache) victim() *Buf {
	var b *Buf
	for {
		cache.clock++
		if cache.clock >= len(cache.buf) {
			cache.clock = 0
		}
		b = &cache.buf[cache.clock]
		if b.inuse == 0 { return b }
	}
	return b
}

func (cache *Cache) lookup(blkno int64) *Buf {
	for i := 0; i < len(cache.buf); i++ {
		if blkno == cache.buf[i].blkno {
			return &cache.buf[i]
		}
	}
	return nil
}

func (cache *Cache) NewBuf() *Buf {
	b := cache.victim()
	cache.dev.NewBlock(b)
	b.inuse++
	return b
}

func (cache *Cache) GetBuf(blkno int64) *Buf {
	b := cache.lookup(blkno)
	if b == nil {
		b = cache.victim()
		b.blkno = blkno
		cache.dev.Fill(b)
	}
	b.inuse++
	return b
}

func (b *Buf) Put() {
	b.cache.dev.Flush(b)
	b.inuse--
}

func main() {
	flag.Parse()
	cachefile := "/tmp/cache"
	if flag.NArg() > 0 {
		cachefile = flag.Arg(0);
	}
	fmt.Printf("cache file = %s\n", cachefile);
	d := OpenDev(cachefile)
	if d == nil {
		fmt.Fprintln(os.Stderr, "open failed", cachefile);
	}
	cache := NewCache(10, d)

// Test 1. Get a buffer, set a value, and Put it back
	b := cache.NewBuf()
	test(b != nil)
	test(b.blkno == 1)
	b.data[0] = 'A'
fmt.Printf("b=%p\n", b)
	b.Put()
	
	cache.Clear()

// Test 2. Get the buffer from test 1 and check it has what we set
// This is not a good test, because it depends on test 1.
// This test did fail to detect that we read the block into a
// separate buffer. Now we have to copies of the buffer in memory.
	b = cache.GetBuf(1)
fmt.Printf("b=%p\n", b)
fmt.Println("b=", b)
	test(b != nil)
	test(b.blkno == 1)
	test(b.data[0] == 'A')
	b.Put()
	test(cache.Audit())

// Test 3. Use up all the buffers without releasing any of them
// Can't really test for running off of end nor can we test for
// infinite loops.
	for i := 0; i < len(cache.buf); i++ {
		b = cache.GetBuf(int64(i))
		b.Put()
	}
	b = cache.GetBuf(int64(len(cache.buf)));
	test(b != nil);

	test(cache.Audit());
}
